scripts/ensure-bundle-deps.ts

```diff
@@ -0,0 +1,146 @@
+  // because `build:bundle` is the slowest part of the build and therefore the one we most want to parallelize with
+  // other slow parts, like `build:types`.)
+  //
+  // In all other scenarios, if the dependencies are missing, we have to build them ourselves - with `build:bundle` at
+  // either level, we're the only thing happening (so no one's going to do it for us), and with package-level `build`,
+  // types and npm assets are being built simultaneously, but only for the package being bundled, not for its
+  // dependencies. Either way, it's on us to fix the problem.
+  //
+  // TODO: This actually *doesn't* work for package-level `build`, not because of a flaw in this logic, but because
+  // `build:rollup` has similar dependency needs (it needs types rather than npm builds). We should do something like
+  // this for that at some point.
+
+  if (isTopLevelBuild && yarnScript === 'build') {
+    let retries = 0;
+
+    console.log('\nSearching for bundle dependencies...');
+
+    while (retries < maxRetries && !checkForBundleDeps(packagesDir, dependencyDirs)) {
+      console.log('Bundle dependencies not found. Trying again in 5 seconds.');
+      retries += 1;
+      await sleep(5000);
+    }
+
+    if (retries === maxRetries) {
+      throw new Error(
+        `\nERROR: \`yarn build:bundle\` (triggered by \`yarn build\`) cannot find its depdendencies, despite waiting ${
+          5 * maxRetries
+        } seconds for the rest of \`yarn build\` to create them. Something is wrong - it shouldn't take that long. Exiting.`,
+      );
+    }
+
+    console.log(`\nFound all bundle dependencies after ${retries} retries. Beginning bundle build...`);
+  }
+
+  // top-level `build:bundle`, package-level `build` and `build:bundle`
+  else {
+    console.warn('\nWARNING: Missing dependencies for bundle build. They will be built before continuing.');
+
+    for (const dependencyDir of dependencyDirs) {
+      console.log(`\nBuilding \`${dependencyDir}\` package...`);
+      run('yarn build:dev', { cwd: `${packagesDir}/${dependencyDir}` });
+    }
+
+    console.log('\nAll dependencies built successfully. Beginning bundle build...');
+  }
+}
+
+/**
+ * See if all of the necessary dependencies exist
+ */
+function checkForBundleDeps(packagesDir: string, dependencyDirs: string[]): boolean {
+  for (const dependencyDir of dependencyDirs) {
+    const depBuildDir = `${packagesDir}/${dependencyDir}/build`;
+    // const depBuildDir = path.resolve(process.cwd(), packagesDir, dependencyDir, 'build');
+
+    // Checking that the directory exists isn't 100% the same as checking that the files themselves exist, of course,
+    // but it's a decent proxy, and much simpler to do than checking  for individual files. (We're arbitrarily checking
+    // for the `cjs` directory here rather than the `esm` directory; they're built together so either one will do.)
+    if (!fs.existsSync(`${depBuildDir}/cjs`) && !fs.existsSync(`${depBuildDir}/npm/cjs`)) {
```

Comment:
we could just as easily check for both `esm` and `cjs` - might be worthwhile just to have anyway.

Reply:
Actually, you're reminding me that it's actually the `esm` that it uses, because we use the default for [this `@rollup/plugin-node-resolve` option](https://github.com/rollup/plugins/tree/master/packages/node-resolve/#mainfields). Added it to the check.